zrml/court/src/migrations.rs

```diff
@@ -14,3 +14,220 @@
+type OldCourtPoolOf<T> = BoundedVec<OldCourtPoolItemOf<T>, <T as Config>::MaxCourtParticipants>;
+
+#[frame_support::storage_alias]
+pub(crate) type CourtPool<T: Config> = StorageValue<Court<T>, OldCourtPoolOf<T>, ValueQuery>;
+
+const COURT_REQUIRED_STORAGE_VERSION: u16 = 2;
+const COURT_NEXT_STORAGE_VERSION: u16 = 3;
+
+pub struct MigrateCourtPoolItems<T>(PhantomData<T>);
+
+impl<T> OnRuntimeUpgrade for MigrateCourtPoolItems<T>
+where
+    T: Config,
+{
+    fn on_runtime_upgrade() -> Weight {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let market_commons_version = StorageVersion::get::<Court<T>>();
+        if market_commons_version != COURT_REQUIRED_STORAGE_VERSION {
+            log::info!(
+                "MigrateCourtPoolItems: court storage version is {:?}, but {:?} is required",
+                market_commons_version,
+                COURT_REQUIRED_STORAGE_VERSION,
+            );
+            return total_weight;
+        }
+        log::info!("MigrateCourtPoolItems: Starting...");
+
+        let res = crate::CourtPool::<T>::translate::<OldCourtPoolOf<T>, _>(|old_pool_opt| {
+            old_pool_opt.map(|mut old_pool| {
+                <CourtPoolOf<T>>::truncate_from(
+                    old_pool
+                        .iter_mut()
+                        .map(|old_pool_item: &mut OldCourtPoolItemOf<T>| CourtPoolItemOf::<T> {
+                            stake: old_pool_item.stake,
+                            court_participant: old_pool_item.court_participant.clone(),
+                            consumed_stake: old_pool_item.consumed_stake,
+                            joined_at: old_pool_item.joined_at,
+                            uneligible_index: old_pool_item
+                                .joined_at
+                                .checked_div(&T::InflationPeriod::get())
+                                // because inflation period is not zero checked_div is safe
+                                // however, if not use the current period index
+                                // 5_184_000 is the block of the inflation reward distribution
+                                // 5_184_000 / 216_000 (blocks per 30 day inflation period) = 24
+                                .unwrap_or(24u64.saturated_into::<BlockNumberOf<T>>()),
+                            // using old_pool_item.stake leads to all joins in period 24
+                            // to be uneligible, which is exactly what we want
+                            // if using zero, all joins of period 24 would be eligible,
+                            // but only joins in 23 waited the full period yet
+                            // to understand the calculation of the eligible stake look into handle_inflation
+                            uneligible_stake: old_pool_item.stake,
+                        })
+                        .collect::<Vec<_>>(),
+                )
+            })
+        });
```

Comment:
Interesting, wondering how to benchmark potentially heavy operations on vectors within a storage here. Clearly we only read and write once, but the actual effort scales with the size of the vector. I think in our case it's irrelevant, but generally, what would be the approach? The only solution I see is benchmarking the mutation of the vector.

Reply:
This is the only solution I see too (benchmarking the migration code in general). I would also always account for the worst case in migration benchmarks, because the migrations are not paid by a user directly as with extrinsics where refunds make sense. In general we never actually accounted for the computation in migrations, only for storage reads and writes, but the clean approach would have been to do exactly this. Apart from the computation made in this mapping here, there is also a bigger encoding and decoding effort for bigger vectors stored in storage. I think Nodle network once made a migration, which exceeded the block limit and led to a chain halt. https://medium.com/nodle-io/latest-parachain-upgrade-what-everyone-should-know-9d5563e97a70

In the article above I see the following solution:

```rust
<Runtime as frame_system::Config>::BlockWeights::get().max_block;
```

This is the true worst weight possible :1st_place_medal: 

It seems using [try-runtime](https://github.com/paritytech/try-runtime-cli/pull/22) allows to see the migration time it took. This means we can watch out for this number if it exceeds the block time.

Reply:
in our case the actual weight was around 0.1% for the try runtime execution